fl4health/utils/warmed_up_module.py

```diff
@@ -0,0 +1,90 @@
+import json
+import os
+from logging import ERROR, INFO
+from typing import Optional
+
+import torch
+from flwr.common.logger import log
+
+
+class WarmedUpModule:
+    """This class is used to load a pretrained model into the current model."""
+
+    def __init__(self, pretrained_model_dir: Optional[str], weights_mapping_dir: Optional[str] = None) -> None:
+        """Initialize the WarmedUpModule with the pretrained model stats and weights mapping dict.
+
+        Args:
+            pretrained_model_dir (Optional[str]): Directory of the pretrained model
+            weights_mapping_dir (Optional[str], optional): Directory of to json file of the weights mapping dict.
+            Defaults to None.
+            If models are not exactly the same, a weights mapping dict is needed to map the weights of the pretrained
+            model to the current model.
+        """
+
+        if pretrained_model_dir is None or not os.path.exists(pretrained_model_dir):
+            log(INFO, "No pretrained model provided")
+            self.pretrained_model_state = None
+        else:
+            log(INFO, f"Loading pretrained model from {pretrained_model_dir}")
+            self.pretrained_model_state = torch.load(pretrained_model_dir).state_dict()
+
+        if weights_mapping_dir is not None:
+            with open(weights_mapping_dir, "r") as file:
+                self.weights_mapping_dict = json.load(file)
+                log(INFO, f"Weights mapping dict: {self.weights_mapping_dict}")
+        else:
+            self.weights_mapping_dict = None
+
+    def get_matching_component(self, key: str) -> Optional[str]:
+        """Get the matching component of the key from the weights mapping dict.
+
+        Args:
+            key (str): Key to be matched in pretrained model.
+
+        Returns:
+            Optional[str]: If no weights mapping dict is provided, returns the key. Otherwise, if the key is in the
+            weights mapping dict, returns the matching component of the key. Otherwise, returns None.
+        """
+
+        if self.weights_mapping_dict is None:
+            return key
+
+        components = key.split(".")
```

Comment:
I am kind of confused why we are splitting the key (corresponding to current model) by dot and reconstructing it by appending component by component and checking each time if the key is present in the weight_mapping_dict. Shouldn't the key be an entry in the dictionary? For example, in the fenda example you provided the config for we would iterate through the keys of the Fenda model. One the keys would be "local_module.conv1" and that could be mapped to "conv1" using the mapping. We dont need to break it apart and reconstruct it piece by piece. 

I may be missing something so please feel free to correct me :) If this is the case, maybe a comment clarifying what is happening here would be helpful

Reply:
Ok, so this function makes module to also support partial mapping. For example now the user can have a mapping such as {"model1":"global_module"}, in that case the model can map anything starting with "model1" to keys starting with "global_module". This is so useful for bigger models which user doesn't one to tell layer by layer mapping.

Reply:
So the idea is that if model gets a key like "model1.conv1.weights" first it separates it and starting from beginning it tries to find a match for "model1", "model1.conv1" and "model1.conv1.weights" in the provided mapping. If it finds a mapping for "model1", it replaces  "model1" in the current key with that mapping.

Reply:
I updated doc string to be more clear, but let me know if it is still confusing.

Reply:
Ah makes a lot of sense! I think the documentation you provided explains it great :)